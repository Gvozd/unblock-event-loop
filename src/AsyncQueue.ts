export class AsyncQueue<D extends any, R extends unknown> {
    constructor(private executor: (data: D) => Promise<R>) {
    }

    private promise: Promise<R> = Promise.resolve(void 0 as R);

    push(data: D): Promise<R> {
        this.promise = this.promise.then(
            () => {
                return this.executor(data);
            },
            () => {
                return this.executor(data);
            }
        );
        return this.promise;
    }
}
