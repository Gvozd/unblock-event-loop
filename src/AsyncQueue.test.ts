import {AsyncQueue} from './AsyncQueue';
import {setTimeout as delay} from 'timers/promises';

describe('EventLoopMonitor', () => {
    const jestConsole = console;
    beforeAll(() => {
        global.console = require('console');
    });
    afterAll(() => {
        global.console = jestConsole;
    });
    it('base', async () => {
        const steps = [];
        var queue = new AsyncQueue(async function(data: number) {
            steps.push({taskStart: data});
            await delay(100);
            steps.push({taskEnd: data});
            return data * 2;
        });

        steps.push('global start');
        var results = [
            queue.push(1),
            queue.push(2),
        ];
        steps.push('global end of push tasks');
        steps.push({tasksExecuted: await Promise.all(results)});
        steps.push('global end');
        expect(steps, 'execution order').toMatchSnapshot();
    });
});
