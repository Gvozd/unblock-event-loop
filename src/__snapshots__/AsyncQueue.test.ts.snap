// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventLoopMonitor base 1`] = `
[
  "global start",
  "global end of push tasks",
  {
    "taskStart": 1,
  },
  {
    "taskEnd": 1,
  },
  {
    "taskStart": 2,
  },
  {
    "taskEnd": 2,
  },
  {
    "tasksExecuted": [
      2,
      4,
    ],
  },
  "global end",
]
`;
